### DISCUSSION TITLE
Update Build Process

### DISCUSSION THREAD

**Discussion by danemorgan** [holger1411/understrap#1358] [Ideas]
### Laravel Mix
https://laravel.com/docs/8.x/mix

### Webpack
https://webpack.js.org/

### Rollup
https://rollupjs.org/guide/en/

### Parcel
https://parceljs.org/

There are a lot of 'how to set this up' blog posts and articles, but there are precious few "this is why you should choose this set up" articles. Ideally, it would be nice to make a decision based on our specific mix. I think that needs to also take into account that at some point we're pretty likely to be adding react to this mix just to stay compatible with the core.

### browserify+gulp vs webpack+gulp vs webpack+npm scripts
https://www.toptal.com/front-end/webpack-browserify-gulp-which-is-better

> Conclusion
>It seems pretty clear to me that Webpack is preferable to Browserify and NPM scripts are preferable to Gulp, although each option has its benefits and drawbacks. Gulp is certainly more expressive and convenient to use than NPM scripts, but you pay the price in all the added abstraction.

**Comment by 0dp** [comment]
I am all on favor of updating the build system.
I just not a big fan of 'just' adding an abstraction layer on top of webpack, to get webpack.

If using a bundler, which i have proposed a few times. Then we actually only need two entries. One for sass and one for js.
All inclusion of files happens in one scss file and one js file as entry points.

Even just leaving the gulp file and add webpack-stream for js files would be a massive improvement.

**Reply by danemorgan** [reply]
I'm kind of in the camp of let's go all-in on a course of action.

I think we all agree we need to update our build process, so let's figure out what works best for our mix of WordPress & Bootstrap. Instead of just picking the shiny new, let's evaluate a few and decide which will serve us best and LTS.

**Reply by 0dp** [reply]
The very wonderful thing about webpack is that we can resolve npm packages directly in the files that needs them. Thus eliminating the need to copy assets.
Thereby simplifying, and cleaning the structure a bit; maybe even put all dev stuff in src 🙈

**Reply by danemorgan** [reply]
Also, as to one CSS file and one JS file, I think that has become an antipattern under HTTPS/HTTP2. There is a lot of evidence that, at this point, smaller files that only load the assets that a given page needs deliver the best speed and resource management. It doesn't deliver the best google score, but Google is behind the curve on this.

**Reply by danemorgan** [reply]
I'm leaning ( pretty hard ) in your direction and thinking Webpack & NPM. But I'm not sure if that's only because after Gulp it's the only one of these I have even a little bit of familiarity with. What if what I don't know I don't know about one of the others means we are missing something wonderful and we're right back here discussing another migration in 12 months?

**Reply by 0dp** [reply]
> Also, as to one CSS file and one JS file, I think that has become an antipattern under HTTPS/HTTP2. There is a lot of evidence that, at this point, smaller files that only load the assets that a given page needs deliver the best speed and resource management. It doesn't deliver the best google score, but Google is behind the curve on this.

I agree but that is where CriticalPath/CriticalCSS would fit in as raised by @axlright somewhere in the issues; and code splitting

**Reply by 0dp** [reply]
> I'm leaning ( pretty hard ) in your direction and thinking Webpack & NPM. But I'm not sure if that's only because after Gulp it's the only one of these I have even a little bit of familiarity with. What if what I don't know I don't know about one of the others means we are missing something wonderful and we're right back here discussing another migration in 12 months?

I am not the one my self to just jump on the band wagon. It's a really deep rabbit hole this whole package management, build bundler tool thing. It's equally daunting as the whole framework thing.

But what we know is, webpack is well documented, loads of example code and modules to pick from.

Let's namedrop a bit:

Parcel, snowpack, fusebox, rollup.js

**Reply by danemorgan** [reply]
@0dp Yes, just so. What I know about those names you dropped are the names. That's where it starts and where it ends. Well, except that a couple of them specifically mention React, and I know we're headed to React land at some point here. I don't think there is any way to avoid that.

**Comment by danemorgan** [comment]
Just adding @0dp's observation elsewhere that this issue is a *prerequisite* to a successful Bootstrap5 implementation. We need the build process sorted before we wrangle that beast.

**Reply by bacoords** [reply]
In that vein, I think we should follow what Bootstrap does - it makes the most sense to use the same build process as Bootstrap itself: https://github.com/twbs/bootstrap/blob/main/package.json . I like that instead of gulp, they pull together a few simple tools like postcss, nodemon, sass, and rollup and just write basic commands in their package.json. Then we can just have a simple `npm start` or something that turns it all on/off.
